File: src/main/java/org/railwaystations/rsapi/adapter/in/web/controller/ProfileController.java
package org.railwaystations.rsapi.adapter.in.web.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.railwaystations.rsapi.adapter.in.web.api.ProfileApi;
import org.railwaystations.rsapi.adapter.in.web.model.ChangePasswordDto;
import org.railwaystations.rsapi.adapter.in.web.model.LicenseDto;
import org.railwaystations.rsapi.adapter.in.web.model.ProfileDto;
import org.railwaystations.rsapi.adapter.in.web.model.RegisterProfileDto;
import org.railwaystations.rsapi.adapter.in.web.model.UpdateProfileDto;
import org.railwaystations.rsapi.core.model.License;
import org.railwaystations.rsapi.core.model.User;
import org.railwaystations.rsapi.core.ports.in.ManageProfileUseCase;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.LocaleResolver;

import java.net.URI;
import java.util.Locale;

import static org.railwaystations.rsapi.adapter.in.web.RequestUtil.getAuthUser;
import static org.railwaystations.rsapi.adapter.in.web.RequestUtil.getRequest;
import static org.railwaystations.rsapi.adapter.in.web.RequestUtil.getUserAgent;

@RestController
@Slf4j
@RequiredArgsConstructor
public class ProfileController implements ProfileApi {

    private final ManageProfileUseCase manageProfileUseCase;

    private final LocaleResolver localeResolver;

    private User toUser(RegisterProfileDto registerProfileDto) {
        return User.builder()
                .name(registerProfileDto.getNickname())
                .email(registerProfileDto.getEmail())
                .url(registerProfileDto.getLink() != null ? registerProfileDto.getLink().toString() : null)
                .ownPhotos(registerProfileDto.getPhotoOwner())
                .anonymous(registerProfileDto.getAnonymous() != null && registerProfileDto.getAnonymous())
                .license(toLicense(registerProfileDto.getLicense()))
                .sendNotifications(registerProfileDto.getSendNotifications() == null || registerProfileDto.getSendNotifications())
                .newPassword(registerProfileDto.getNewPassword())
                .locale(localeResolver.resolveLocale(getRequest()))
                .build();
    }

    private License toLicense(LicenseDto license) {
        if (license == null) {
            return License.UNKNOWN;
        }
        return switch (license) {
            case CC0_1_0_UNIVERSELL_CC0_1_0_, CC0 -> License.CC0_10;
            case CC_BY_SA_4_0, CC4 -> License.CC_BY_SA_40;
            case UNKNOWN -> License.UNKNOWN;
        };
    }

    private ProfileDto toProfileDto(User user) {
        return new ProfileDto(user.getName(), toLicenseDto(user.getLicense()), user.isOwnPhotos())
                .admin(user.isAdmin())
                .email(user.getEmail())
                .anonymous(user.isAnonymous())
                .emailVerified(user.isEmailVerified())
                .sendNotifications(user.isSendNotifications())
                .link(user.getUrl() != null ? URI.create(user.getUrl()) : null);
    }

    private LicenseDto toLicenseDto(License license) {
        if (license == null) {
            return LicenseDto.UNKNOWN;
        }
        return switch (license) {
            case CC0_10 -> LicenseDto.CC0_1_0_UNIVERSELL_CC0_1_0_;
            case CC_BY_SA_40 -> LicenseDto.CC_BY_SA_4_0;
            default -> LicenseDto.UNKNOWN;
        };
    }

    private User toUser(UpdateProfileDto updateProfileDto) {
        Locale locale = localeResolver.resolveLocale(getRequest());
        log.info("User locale {}", locale);
        return User.builder()
                .name(updateProfileDto.getNickname())
                .email(updateProfileDto.getEmail())
                .url(updateProfileDto.getLink() != null ? updateProfileDto.getLink().toString() : null)
                .ownPhotos(updateProfileDto.getPhotoOwner())
                .anonymous(updateProfileDto.getAnonymous() != null && updateProfileDto.getAnonymous())
                .license(toLicense(updateProfileDto.getLicense()))
                .sendNotifications(updateProfileDto.getSendNotifications() == null || updateProfileDto.getSendNotifications())
                .locale(locale)
                .build();
    }

    @PreAuthorize("isAuthenticated()")
    @Override
    public ResponseEntity<Void> changePasswordPost(String newPassword, ChangePasswordDto changePasswordDto) {
        manageProfileUseCase.changePassword(getAuthUser().getUser(), changePasswordDto != null ? changePasswordDto.getNewPassword() : newPassword);
        return ResponseEntity.ok().build();
    }

    @Override
    public ResponseEntity<String> emailVerificationTokenGet(String token) {
        return manageProfileUseCase.emailVerification(token)
                .map(u -> new ResponseEntity<>("Email successfully verified!", HttpStatus.OK))
                .orElse(ResponseEntity.notFound().build());
    }

    @PreAuthorize("isAuthenticated()")
    @Override
    public ResponseEntity<Void> myProfileDelete() {
        manageProfileUseCase.deleteProfile(getAuthUser().getUser(), getUserAgent());
        return ResponseEntity.noContent().build();
    }

    @PreAuthorize("isAuthenticated()")
    @Override
    public ResponseEntity<ProfileDto> myProfileGet() {
        User user = getAuthUser().getUser();
        log.info("Get profile for '{}'", user.getEmail());
        return ResponseEntity.ok(toProfileDto(user));
    }

    @PreAuthorize("isAuthenticated()")
    @Override
    public ResponseEntity<Void> myProfilePost(UpdateProfileDto profile) {
        manageProfileUseCase.updateProfile(getAuthUser().getUser(), toUser(profile), getUserAgent());
        return ResponseEntity.ok().build();
    }

    @Override
    public ResponseEntity<Void> registrationPost(RegisterProfileDto registration) {
        manageProfileUseCase.register(toUser(registration), getUserAgent());
        return ResponseEntity.accepted().build();
    }

    @PreAuthorize("isAuthenticated()")
    @Override
    public ResponseEntity<Void> resendEmailVerificationPost() {
        manageProfileUseCase.resendEmailVerification(getAuthUser().getUser());
        return ResponseEntity.ok().build();
    }

    @Override
    public ResponseEntity<Void> resetPasswordPost(String nameOrEmail) {
        manageProfileUseCase.resetPassword(nameOrEmail, getUserAgent());
        return ResponseEntity.accepted().build();
    }

}

File: adapter/src/main/kotlin/org/railwaystations/rsapi/adapter/web/controller/ProfileController.kt
package org.railwaystations.rsapi.adapter.web.controller

import org.railwaystations.rsapi.adapter.web.RequestUtil
import org.railwaystations.rsapi.adapter.web.api.ProfileApi
import org.railwaystations.rsapi.adapter.web.model.ChangePasswordDto
import org.railwaystations.rsapi.adapter.web.model.LicenseDto
import org.railwaystations.rsapi.adapter.web.model.ProfileDto
import org.railwaystations.rsapi.adapter.web.model.UpdateProfileDto
import org.railwaystations.rsapi.core.model.License
import org.railwaystations.rsapi.core.model.User
import org.railwaystations.rsapi.core.ports.inbound.ManageProfileUseCase
import org.railwaystations.rsapi.core.utils.Logger
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.servlet.LocaleResolver
import java.net.URI
import java.util.*

@RestController
class ProfileController(
    private val manageProfileUseCase: ManageProfileUseCase,
    private val localeResolver: LocaleResolver,
    private val requestUtil: RequestUtil,
) : ProfileApi {

    private val log by Logger()

    @PreAuthorize("isAuthenticated()")
    override fun postChangePassword(authorization: String, changePasswordDto: ChangePasswordDto): ResponseEntity<Unit> {
        manageProfileUseCase.changePassword(requestUtil.authUser.user, changePasswordDto.newPassword)
        return ResponseEntity.ok().build()
    }

    override fun getEmailVerification(token: String): ResponseEntity<String> {
        return manageProfileUseCase.emailVerification(token)
            ?.let { _ ->
                ResponseEntity(
                    "Email successfully verified!",
                    HttpStatus.OK
                )
            }
            ?: ResponseEntity.notFound().build()
    }

    @PreAuthorize("isAuthenticated()")
    override fun deleteMyProfile(authorization: String): ResponseEntity<Unit> {
        manageProfileUseCase.deleteProfile(requestUtil.authUser.user, requestUtil.userAgent)
        return ResponseEntity.noContent().build()
    }

    @PreAuthorize("isAuthenticated()")
    override fun getMyProfile(authorization: String): ResponseEntity<ProfileDto> {
        val user = requestUtil.authUser.user
        log.info("Get profile for '{}'", user.email)
        return ResponseEntity.ok(user.toDto())
    }

    @PreAuthorize("isAuthenticated()")
    override fun postMyProfile(authorization: String, profile: UpdateProfileDto): ResponseEntity<Unit> {
        val locale = localeResolver.resolveLocale(requestUtil.request)
        log.info("User locale {}", locale)
        manageProfileUseCase.updateProfile(requestUtil.authUser.user, profile.toDomain(locale), requestUtil.userAgent)
        return ResponseEntity.ok().build()
    }

    @PreAuthorize("isAuthenticated()")
    override fun postResendEmailVerification(authorization: String): ResponseEntity<Unit> {
        manageProfileUseCase.resendEmailVerification(requestUtil.authUser.user)
        return ResponseEntity.ok().build()
    }

}

private fun UpdateProfileDto.toDomain(locale: Locale) = User(
    name = nickname,
    email = email,
    url = link?.toString(),
    ownPhotos = photoOwner ?: false,
    anonymous = anonymous ?: false,
    license = license.toDomain(),
    sendNotifications = sendNotifications ?: true,
    locale = locale,
)

private fun User.toDto() = ProfileDto(
    nickname = name,
    license = license.toDto(),
    photoOwner = ownPhotos,
    email = email,
    link = url?.let { URI.create(it) },
    anonymous = anonymous,
    admin = admin,
    emailVerified = isEmailVerified,
    sendNotifications = sendNotifications,
)

private fun License?.toDto() = when (this) {
    License.CC0_10
    -> LicenseDto.CC0_1_PERIOD0_UNIVERSELL_LEFT_PARENTHESIS_CC0_1_PERIOD0_RIGHT_PARENTHESIS

    License.CC_BY_SA_40
    -> LicenseDto.CC_BY_MINUS_SA_4_PERIOD0

    else -> LicenseDto.UNKNOWN
}

fun LicenseDto?.toDomain() = when (this) {
    LicenseDto.CC0_1_PERIOD0_UNIVERSELL_LEFT_PARENTHESIS_CC0_1_PERIOD0_RIGHT_PARENTHESIS, LicenseDto.CC0
    -> License.CC0_10

    LicenseDto.CC_BY_MINUS_SA_4_PERIOD0, LicenseDto.CC4
    -> License.CC_BY_SA_40

    LicenseDto.UNKNOWN, null -> License.UNKNOWN
}

